test/runner/run.js

```diff
@@ -0,0 +1,306 @@
+import chalk from "chalk";
+import { asyncExitHook, gracefulExit } from "exit-hook";
+import { getLatestBrowser } from "./browserstack/api.js";
+import { buildBrowserFromString } from "./browserstack/buildBrowserFromString.js";
+import { localTunnel } from "./browserstack/local.js";
+import { reportEnd, reportTest } from "./reporter.js";
+import {
+	cleanupAllWorkers,
+	cleanupWorker,
+	debugWorker,
+	retryTest,
+	touchWorker
+} from "./browserstack/workers.js";
+import { createTestServer } from "./createTestServer.js";
+import { buildTestUrl } from "./lib/buildTestUrl.js";
+import { generateHash, printModuleHashes } from "./lib/generateHash.js";
+import { getBrowserString } from "./lib/getBrowserString.js";
+import { addRun, runFullQueue } from "./queue.js";
+import { cleanupAllJSDOM, cleanupJSDOM } from "./jsdom.js";
+
+/**
+ * Run modules in parallel in different browser instances.
+ */
+export async function run( {
+	browsers,
+	browserstack,
+	concurrency,
+	debug,
+	esm,
+	headless,
+	isolate = true,
+	modules = [],
+	retries = 3,
+	verbose
+} = {} ) {
+	if ( !browsers || !browsers.length ) {
+		browsers = [ "chrome" ];
+	}
+	if ( headless && debug ) {
+		throw new Error(
+			"Cannot run in headless mode and debug mode at the same time."
+		);
+	}
+
+	if ( verbose ) {
+		console.log( browserstack ? "Running in BrowserStack." : "Running locally." );
+	}
+
+	const errorMessages = [];
+	const pendingErrors = {};
+
+	// Convert browsers to browser objects
+	browsers = browsers.map( ( b ) => ( { browser: b } ) );
```

Comment:
Same remark as somewhere above - I'd prefer a different variable when the shape of the data changes; even if we're not using TypeScript, it makes it easier to reason about the code. Plus, we have an option to one day experiment with an JSDoc-based TS type checking which would require such a separation.